src/compiler/checker.ts

```diff
@@ -22507,9 +22512,23 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
             }
             else if (sourceFlags & TypeFlags.Index) {
-                if (result = isRelatedTo(keyofConstraintType, target, RecursionFlags.Source, reportErrors)) {
+                const isDeferredMappedIndex = shouldDeferIndexType((source as IndexType).type, (source as IndexType).indexFlags) && getObjectFlags((source as IndexType).type) & ObjectFlags.Mapped;
+                if (result = isRelatedTo(keyofConstraintType, target, RecursionFlags.Source, reportErrors && !isDeferredMappedIndex)) {

```

Comment:
I don't know that it matters a lot, but shouldn't `!isDeferredMappedIndex` come first to avoid calling `isRelatedTo` when the source is in fact a deferred mapped index? Or is it still useful to call `isRelatedTo` in this case?

Reply:
That's not after the call, it's *in* the call, modifying the `reportErrors` parameter to it.